File: tests/fixtures/sample_project_ts/src/billing/invoices.ts
// Invoice generation and delivery.

import { getConnection } from "../db/connection";
import { getUser } from "../users/crud";

/** Write an invoice row and return it. */
export function createInvoice(
  userId: number,
  amountCents: number,
  chargeId: string,
  description: string,
): Record<string, unknown> {
  const conn = getConnection();
  const result = conn.execute(
    "INSERT INTO invoices (user_id, amount_cents, charge_id, description, created_at) VALUES (?, ?, ?, ?, ?)",
    [userId, amountCents, chargeId, description, Math.floor(Date.now() / 1000)],
  );
  return {
    id: result.lastId,
    userId,
    amountCents,
    chargeId,
    description,
  };
}

/**
 * Render an invoice and hand it to the email transport.
 *
 * In production this would call the mail service; here it just marks
 * the invoice as sent.
 */
export function sendInvoiceEmail(invoiceId: number): void {
  const conn = getConnection();
  const { rows } = conn.execute(
    "SELECT user_id, amount_cents, description FROM invoices WHERE id = ?",
    [invoiceId],
  );
  if (rows.length === 0) {
    throw new InvoiceNotFoundError(`invoice ${invoiceId} not found`);
  }
  const row = rows[0];
  const user = getUser(row[0] as number);
  if (user === null) {
    throw new InvoiceNotFoundError(`invoice ${invoiceId} user missing`);
  }
  renderAndQueue(user.email, invoiceId, row[1] as number, row[2] as string);
  conn.execute("UPDATE invoices SET sent_at = ? WHERE id = ?", [
    Math.floor(Date.now() / 1000),
    invoiceId,
  ]);
}

/** Return all invoices for a user, newest first. */
export function listUserInvoices(userId: number): Record<string, unknown>[] {
  const conn = getConnection();
  const { rows } = conn.execute(
    "SELECT id, amount_cents, charge_id, description, created_at, sent_at FROM invoices WHERE user_id = ? ORDER BY created_at DESC",
    [userId],
  );
  return rows.map((r) => ({
    id: r[0],
    amountCents: r[1],
    chargeId: r[2],
    description: r[3],
    createdAt: r[4],
    sentAt: r[5],
  }));
}

/** Stand-in for the real mail-queue call. */
function renderAndQueue(
  email: string,
  invoiceId: number,
  amount: number,
  description: string,
): void {
  void [email, invoiceId, amount, description];
}

/** Raised when referencing a missing invoice. */
export class InvoiceNotFoundError extends Error {}

File: neuralmind/demo_data/sample_project/billing/invoices.py
"""Invoice generation and delivery."""

from __future__ import annotations

import time

from ..db.connection import get_connection
from ..users.crud import get_user


def create_invoice(user_id: int, amount_cents: int, charge_id: str, description: str) -> dict:
    """Write an invoice row and return it."""
    conn = get_connection()
    cursor = conn.execute(
        "INSERT INTO invoices (user_id, amount_cents, charge_id, description, created_at) "
        "VALUES (?, ?, ?, ?, ?) RETURNING id",
        (user_id, amount_cents, charge_id, description, int(time.time())),
    )
    return {
        "id": cursor.fetchone()[0],
        "user_id": user_id,
        "amount_cents": amount_cents,
        "charge_id": charge_id,
        "description": description,
    }


def send_invoice_email(invoice_id: int) -> None:
    """Render an invoice and hand it to the email transport.

    In production this would call the mail service; here it just marks
    the invoice as sent.
    """
    conn = get_connection()
    row = conn.execute(
        "SELECT user_id, amount_cents, description FROM invoices WHERE id = ?",
        (invoice_id,),
    ).fetchone()
    if row is None:
        raise InvoiceNotFoundError(f"invoice {invoice_id} not found")

    user = get_user(row[0])
    if user is None:
        raise InvoiceNotFoundError(f"invoice {invoice_id} user missing")

    _render_and_queue(user.email, invoice_id, row[1], row[2])

    conn.execute(
        "UPDATE invoices SET sent_at = ? WHERE id = ?",
        (int(time.time()), invoice_id),
    )


def list_user_invoices(user_id: int) -> list[dict]:
    """Return all invoices for a user, newest first."""
    conn = get_connection()
    rows = conn.execute(
        "SELECT id, amount_cents, charge_id, description, created_at, sent_at "
        "FROM invoices WHERE user_id = ? ORDER BY created_at DESC",
        (user_id,),
    ).fetchall()
    return [
        {
            "id": r[0],
            "amount_cents": r[1],
            "charge_id": r[2],
            "description": r[3],
            "created_at": r[4],
            "sent_at": r[5],
        }
        for r in rows
    ]


def _render_and_queue(email: str, invoice_id: int, amount: int, description: str) -> None:
    """Stand-in for the real mail-queue call."""
    _ = (email, invoice_id, amount, description)


class InvoiceNotFoundError(Exception):
    """Raised when referencing a missing invoice."""
